plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -480,10 +489,40 @@ private Optional<Table> loadTable(HiveTableName hiveTableName)
         return delegate.getTable(hiveTableName.getDatabaseName(), hiveTableName.getTableName());
     }
 
+    /**
+     * The method will cache and return columns specified in the {@link Table#getDataColumns()}
+     * but may return more if other columns are already cached.
+     */
     @Override
     public PartitionStatistics getTableStatistics(Table table)
     {
-        return get(tableStatisticsCache, hiveTableName(table.getDatabaseName(), table.getTableName()), () -> delegate.getTableStatistics(table));
+        HiveTableName tableName = hiveTableName(table.getDatabaseName(), table.getTableName());
+        PartitionStatistics statistics = get(tableStatisticsCache, tableName, () -> delegate.getTableStatistics(table));
+        Set<String> dataColumns = table.getDataColumns().stream().map(Column::getName).collect(toImmutableSet());
+        if (statistics.getColumnStatistics().keySet().containsAll(dataColumns)) {
+            // all requested column stats are already cached
+            return statistics;
+        }
+
+        // some columns are missing in the cache, lets retrieve the missing stats
+        SetView<String> missingColumns = difference(dataColumns, statistics.getColumnStatistics().keySet());
+        Map<String, Column> columnNameToColumn = Maps.uniqueIndex(table.getDataColumns(), Column::getName);
+        Table tableWithOnlyMissingColumns = Table.builder(table)
+                .setDataColumns(missingColumns.stream().map(columnNameToColumn::get).collect(toImmutableList()))
+                .build();
+        PartitionStatistics missingColumnStatistics = delegate.getTableStatistics(tableWithOnlyMissingColumns);
+
+        // merge missing stats into the cached stats
+        return tableStatisticsCache.asMap().compute(
+                tableName,
+                (ignored, oldStats) -> oldStats != null
+                        ? new PartitionStatistics(
+                        missingColumnStatistics.getBasicStatistics(),
+                        ImmutableMap.<String, HiveColumnStatistics>builder()
```

Comment:
nit: looks like IDE gets the indentation wrong here, maybe add a tab manually, parameters usually don't start from behind the invocation on previous line